lib/Dialect/Comb/CombFolds.cpp

```diff
@@ -962,8 +973,9 @@ LogicalResult AndOp::canonicalize(AndOp op, PatternRewriter &rewriter) {
     return success();
 
   // and(..., x, ..., x) -> and(..., x, ...) -- idempotent
+  // and(..., x, and(..., x, ...)) -> and(..., and(..., x, ...)) -- idempotent
   // Trivial and(x), and(x, x) cases are handled by [AndOp::fold] above.
-  if (size > 2 && canonicalizeIdempotentInputs(op, rewriter))
+  if (size > 1 && canonicalizeIdempotentInputs(op, rewriter))
     return success();
```

Comment:
> Could it make sense to tweak your implementation to not inline nested and ops, but to filter out any and operands that were already covered by a nested one? For example, by doing a pass over all nested and ops, adding their operands into a set, and then doing a second pass to remove all operands from the main and op that are already in that set.

I think that makes a whole lot of sense, and that is what I attempted to do in this PR. I had some difficulties wording the pull request text, and the attempted explanation turned out very poor.

> You could also use that set of already covered operands to prevent re-visiting operations through the worklist if you've already seen them, which would prevent the canonicalizer from looping infinitely if the IR has a combinational loop in it.

I had not thought about combinational loops, oops!

My initial thought was to do

```diff
diff --git a/lib/Dialect/Comb/CombFolds.cpp b/lib/Dialect/Comb/CombFolds.cpp
index 0cb0464ad..cda924a0b 100644
--- a/lib/Dialect/Comb/CombFolds.cpp
+++ b/lib/Dialect/Comb/CombFolds.cpp
@@ -936,10 +936,14 @@ template <typename Op>
 static bool canonicalizeIdempotentInputs(Op op, PatternRewriter &rewriter) {
   auto inputs = op.getInputs();
 
-  llvm::SmallSetVector<Value, 8> alreadyCovered;
   llvm::SmallVector<Value, 8> worklist(inputs.begin(), inputs.end());
+  llvm::SmallSetVector<Value, 8> alreadyCovered;
+  alreadyCovered.insert(op);
+
   while (!worklist.empty()) {
     auto element = worklist.pop_back_val();
+    if (alreadyCovered.contains(element))
+      continue;
 
     if (auto idempotentOp = element.getDefiningOp<Op>()) {
       alreadyCovered.insert(idempotentOp.getInputs().begin(),
```

but that prevents us from recursing to `%0` in when canonicalizing `%2` in
```
hw.module @Baz(in %a : i2, in %b : i2, in %c : i2, out o1: i2, out o2: i2, out o3: i2) {
  %0 = comb.and bin %a, %b : i2
  %1 = comb.and bin %0, %c : i2
  %2 = comb.and bin %1, %a : i2
  hw.output %0, %1, %2: i2, i2, i2
}
```
resulting in
```
module {
  hw.module @Baz(in %a : i2, in %b : i2, in %c : i2, out o1 : i2, out o2 : i2, out o3 : i2) {
    %0 = comb.and bin %a, %b : i2
    %1 = comb.and bin %0, %c : i2
    %2 = comb.and bin %1, %a : i2
    hw.output %0, %1, %2 : i2, i2, i2
  }
}
```
instead of
```
module {
  hw.module @Baz(in %a : i2, in %b : i2, in %c : i2, out o1 : i2, out o2 : i2, out o3 : i2) {
    %0 = comb.and bin %a, %b : i2
    %1 = comb.and bin %0, %c : i2
    hw.output %0, %1, %1 : i2, i2, i2
  }
}
```

I've pushed a proposal to 29b793b0eeda292f4f2ac3bb79d4dfdd236a47da, but it didn't turn out as pretty as I would like, and some advice on cleaning it up would be greatly appreciated.